src/jsep.js

```diff
@@ -2,257 +2,524 @@
+	 */
+	addIdentifierChar(char) {
+		this.additional_identifier_chars.add(char);
+		return this;
+	}
+
+	/**
+	 * @method addLiteral
+	 * @param {string} literal_name The name of the literal to add
+	 * @param {*} literal_value The value of the literal
+	 * @return this
+	 */
+	addLiteral(literal_name, literal_value) {
+		this.literals[literal_name] = literal_value;
+		return this;
+	}
+
+	/**
+	 * @method removeUnaryOp
+	 * @param {string} op_name The name of the unary op to remove
+	 * @return this
+	 */
+	removeUnaryOp(op_name) {
+		delete this.unary_ops[op_name];
+		if (op_name.length === this.max_unop_len) {
+			this.max_unop_len = JsepConfig.getMaxKeyLen(this.unary_ops);
+		}
+		return this;
+	}
+
+	/**
+	 * @method removeAllUnaryOps
+	 * @return this
+	 */
+	removeAllUnaryOps() {
+		this.unary_ops = {};
+		this.max_unop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeIdentifierChar
+	 * @param {string} char The additional character to stop treating as a valid part of an identifier
+	 * @return this
+	 */
+	removeIdentifierChar(char) {
+		this.additional_identifier_chars.delete(char);
+		return this;
+	}
+
+	/**
+	 * @method removeBinaryOp
+	 * @param {string} op_name The name of the binary op to remove
+	 * @return this
+	 */
+	removeBinaryOp(op_name) {
+		delete this.binary_ops[op_name];
+
+		if (op_name.length === this.max_binop_len) {
+			this.max_binop_len = JsepConfig.getMaxKeyLen(this.binary_ops);
+		}
+
+		return this;
+	}
+
+	/**
+	 * @method removeAllBinaryOps
+	 * @return this
+	 */
+	removeAllBinaryOps() {
+		this.binary_ops = {};
+		this.max_binop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeLiteral
+	 * @param {string} literal_name The name of the literal to remove
+	 * @return this
+	 */
+	removeLiteral(literal_name) {
+		delete this.literals[literal_name];
+		return this;
+	}
+
+	/**
+	 * @method removeAllLiterals
+	 * @return this
+	 */
+	removeAllLiterals() {
+		this.literals = {};
+
+		return this;
+	}
+}
+
+
+// JavaScript Expression Parser
+export class Jsep {
+	static get version() {
+		// To be filled in by the template
+		return '<%= version %>';
+	}
+
+	/**
+	 * @returns {(pos: number) => string}
+	 */
+	get charAtFunc() {
+		return this.expr.charAt;
+	}
+
+	/**
+	 * @returns {(index: number) => number}
+	 */
+	get charCodeAtFunc() {
+		return this.expr.charCodeAt;
+	}
+
+	get _exprI() {
+		return this.exprI(this.index);
+	}
+
+	get _exprICode() {
+		return this.exprICode(this.index);
 	};
-	let exprICode = function(i) {
-		return charCodeAtFunc.call(expr, i);
+
+
+	/**
+	 * @param {string} expr a string with the passed in express
+	 * @param {JsepConfig} [config]
+	 * @returns Jsep
+	 */
+	constructor(expr, config = JsepConfig.instance) {
+		// `index` stores the character number we are currently at while `length` is a constant
+		// All of the gobbles below will modify `index` as we move along
+		this.expr = expr;
+		this.config = config;
+		this.index = 0;
+		this.length = expr.length;
```

Comment:
Maybe this should be a getter? Though that might hinder performance.

Reply:
I just ran a performance benchmark and there doesn't seem to be any performance difference between saving that local variable versus using `this.expr.length`. A getter is about 60% slower though. Mostly I just kept it to try not to change the existing code too much